eva_submission/biosamples_submission.py

```diff
@@ -170,57 +172,106 @@ def communicator_attributes(self):
         return {'webinSubmissionAccountId': self.username}
 
 
+class BioSamplesSubmitter(AppLogger):
 
-class BSDSubmitter(AppLogger):
+    valid_actions = ('create', 'overwrite', 'curate', 'derive')
 
-    def __init__(self, communicator):
-        self.communicator = communicator
+    def __init__(self, communicators, submit_type=('create',), allow_removal=False):
+        assert len(communicators) > 0, 'Specify at least one communicator object to BioSamplesSubmitter'
+        assert set(submit_type) <= set(self.valid_actions), f'all actions must be in {self.valid_actions}'
+        self.default_communicator = communicators[0]
+        self.communicators = communicators
         self.sample_name_to_accession = {}
-
-    def should_create(self, sample):
-        return 'accession' not in sample
-
-    def should_update(self, sample):
-        return 'accession' in sample and 'name' in sample
-
-    def should_retrieve(self, sample):
-        return 'accession' in sample and 'name' not in sample
+        self.submit_type = submit_type
+        self.allow_removal = allow_removal
+
+    @lru_cache
+    def _get_existing_sample(self, accession):
+        return self.default_communicator.follows_link('samples', method='GET', join_url=accession)
+
+    def can_create(self, sample):
+        return 'create' in self.submit_type and 'accession' not in sample
+
+    def can_overwrite(self, sample):
+        """ We should overwrite a samples when it is owned by a domain supported current uploader"""
+        return 'overwrite' in self.submit_type and \
+            'accession' in sample and \
+            self._get_sample_owner(sample)
+
+    def can_curate(self, sample):
+        """ We can curate a samples if it has an existing accessionr"""
+        return 'curate' in self.submit_type and 'accession' in sample
+
+    def can_derive(self, sample):
+        return 'derive' in self.submit_type and 'accession' in sample
+
+    def _get_sample_owner(self, sample):
```

Comment:
If this is only used to check whether some EVA account owns a sample, would prefer it to return a boolean and be named as such.

Reply:
You're right, It is used to get the EVA account that own this sample. I don't think a boolean would work though. But I agree that the name is poorly chosen.